File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = 0;
    
    sort(times.begin(), times.end());
    
    long long left = 1;
    long long right = times[times.size() - 1] * (long long)n;
    long long mid;
    long long cnt;
    
    while(left <= right) {
        mid = (left + right) / 2;
        
        cnt = 0;
        for (int i = 0; i < times.size(); i++) {
            cnt += mid / times[i];
        }
        
        if (cnt < n) {
            left = mid + 1;
        } else {
            right = mid - 1; 
            answer = mid;
        }
    }
    

    return answer;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.js
function solution(n, times) {
    times.sort((a, b) => {
        return a - b;
    });
    
    let left = 0;
    let right = times[times.length - 1] * n; // 최대 소요 시간
    let answer = Infinity;
    
    while(left <= right) {
        let mid = Math.floor((left + right) / 2);
        
        let cnt = 0;
        times.forEach((time) => {
            cnt += Math.floor(mid / time);
        });
        
        if (cnt >= n) {
            answer = Math.min(answer, mid)
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }
    
    return answer;
}